File: src/cwlint/diff.py
"""Diff two contract JSON objects and render a human-readable markdown report."""

from __future__ import annotations

from typing import Any


def diff_specs(old: dict[str, Any], new: dict[str, Any]) -> dict[str, Any]:
    """Produce a structured diff. Categories:
    - meta_changed: top-level metadata (versions, extracted_at)
    - sets_changed: per-named-set additions/removals
    - other_changed: other top-level keys whose value differs
    """
    meta_keys = {"claude_app_version", "operon_core_version", "claude_cli_version", "extracted_at"}
    meta_changed: dict[str, dict[str, Any]] = {}
    sets_changed: dict[str, dict[str, Any]] = {}
    other_changed: dict[str, Any] = {}

    for key in meta_keys:
        if old.get(key) != new.get(key):
            meta_changed[key] = {"old": old.get(key), "new": new.get(key)}

    # Collect every named-string set in both old and new for diff
    for path in _named_set_paths(old) | _named_set_paths(new):
        old_names = set(_get_at(old, path).get("names", []) if _get_at(old, path) else [])
        new_names = set(_get_at(new, path).get("names", []) if _get_at(new, path) else [])
        added = sorted(new_names - old_names)
        removed = sorted(old_names - new_names)
        if added or removed:
            sets_changed[".".join(path)] = {"added": added, "removed": removed}

    # Other top-level keys (not yet covered by meta or named sets)
    for key in set(old) | set(new):
        if key in meta_keys:
            continue
        if old.get(key) != new.get(key):
            # Skip keys we've already accounted for via named-set walking.
            if any(p[0] == key for p in _named_set_paths(old) | _named_set_paths(new)):
                continue
            other_changed[key] = "value differs (no named-set walk)"

    return {
        "meta_changed": meta_changed,
        "sets_changed": sets_changed,
        "other_changed": other_changed,
    }


def render_markdown_diff(diff: dict[str, Any], old_version: str, new_version: str) -> str:
    """Render a human-readable PR-body markdown from the structured diff."""
    lines: list[str] = []
    lines.append(f"# Cowork contract: {old_version} → {new_version}\n")

    if diff["meta_changed"]:
        lines.append("## Metadata\n")
        for key, change in diff["meta_changed"].items():
            lines.append(f"- `{key}`: `{change['old']}` → `{change['new']}`")
        lines.append("")

    if diff["sets_changed"]:
        lines.append("## Named-set changes\n")
        for set_path, change in sorted(diff["sets_changed"].items()):
            added = change["added"]
            removed = change["removed"]
            lines.append(f"### `{set_path}`\n")
            if added:
                lines.append(f"- **Added** ({len(added)}): {', '.join(f'`{n}`' for n in added)}")
            if removed:
                rem_str = ", ".join(f"`{n}`" for n in removed)
                lines.append(f"- **Removed** ({len(removed)}): {rem_str}")
            lines.append("")

    if diff["other_changed"]:
        lines.append("## Other top-level changes\n")
        for key in sorted(diff["other_changed"]):
            lines.append(f"- `{key}`: value differs")
        lines.append("")

    if not (diff["meta_changed"] or diff["sets_changed"] or diff["other_changed"]):
        lines.append("_No differences detected._\n")

    return "\n".join(lines)


def _named_set_paths(obj: Any, path: tuple[str, ...] = ()) -> set[tuple[str, ...]]:
    out: set[tuple[str, ...]] = set()
    if isinstance(obj, dict):
        if "names" in obj and isinstance(obj["names"], list) and path:
            out.add(path)
        for key, value in obj.items():
            out |= _named_set_paths(value, (*path, str(key)))
    return out


def _get_at(obj: Any, path: tuple[str, ...]) -> Any:
    cur = obj
    for part in path:
        if not isinstance(cur, dict) or part not in cur:
            return None
        cur = cur[part]
    return cur

File: src/diff.ts
/**
 * Diff two contract JSON objects and render a human-readable markdown report.
 *
 * Port of `_legacy/python/cwlint/diff.py`. Three categories:
 *   - meta_changed:  fixed top-level metadata (versions, extracted_at)
 *   - sets_changed:  per-named-set additions/removals, where a "named set" is
 *                    any nested `{ "names": [...] }` object the walker finds
 *   - other_changed: any other top-level key whose JSON value differs and
 *                    isn't already accounted for via named-set walking
 *
 * `renderMarkdownDiff()` produces a PR-body-ready markdown summary; if every
 * category is empty it emits a single `_No differences detected._` line.
 */

export interface SpecDiff {
  meta_changed: Record<string, { old: unknown; new: unknown }>;
  sets_changed: Record<string, { added: string[]; removed: string[] }>;
  other_changed: Record<string, "value differs (no named-set walk)">;
}

const META_KEYS = new Set([
  "claude_app_version",
  "operon_core_version",
  "claude_cli_version",
  "extracted_at",
]);

export function diffSpecs(old: Record<string, unknown>, next: Record<string, unknown>): SpecDiff {
  const metaChanged: SpecDiff["meta_changed"] = {};
  const setsChanged: SpecDiff["sets_changed"] = {};
  const otherChanged: SpecDiff["other_changed"] = {};

  for (const key of META_KEYS) {
    if (!deepEqual(old[key], next[key])) {
      metaChanged[key] = { old: old[key], new: next[key] };
    }
  }

  // Collect every named-string set in both old and new for diff
  const oldPaths = namedSetPaths(old);
  const newPaths = namedSetPaths(next);
  const allPaths = new Map<string, string[]>();
  for (const [k, v] of oldPaths) {
    allPaths.set(k, v);
  }
  for (const [k, v] of newPaths) {
    allPaths.set(k, v);
  }

  for (const path of allPaths.values()) {
    const oldNode = getAt(old, path);
    const newNode = getAt(next, path);
    const oldNames = new Set(extractNames(oldNode));
    const newNames = new Set(extractNames(newNode));
    const added = [...newNames].filter((n) => !oldNames.has(n)).sort();
    const removed = [...oldNames].filter((n) => !newNames.has(n)).sort();
    if (added.length > 0 || removed.length > 0) {
      setsChanged[path.join(".")] = { added, removed };
    }
  }

  // Other top-level keys (not yet covered by meta or named sets)
  const namedTopLevels = new Set<string>();
  for (const path of allPaths.values()) {
    const head = path[0];
    if (head !== undefined) {
      namedTopLevels.add(head);
    }
  }
  const allKeys = new Set([...Object.keys(old), ...Object.keys(next)]);
  for (const key of allKeys) {
    if (META_KEYS.has(key)) continue;
    if (deepEqual(old[key], next[key])) continue;
    if (namedTopLevels.has(key)) continue;
    otherChanged[key] = "value differs (no named-set walk)";
  }

  return {
    meta_changed: metaChanged,
    sets_changed: setsChanged,
    other_changed: otherChanged,
  };
}

export function renderMarkdownDiff(diff: SpecDiff, oldVersion: string, newVersion: string): string {
  const lines: string[] = [];
  lines.push(`# Cowork contract: ${oldVersion} → ${newVersion}\n`);

  const metaKeys = Object.keys(diff.meta_changed);
  if (metaKeys.length > 0) {
    lines.push("## Metadata\n");
    for (const key of metaKeys) {
      const change = diff.meta_changed[key];
      lines.push(`- \`${key}\`: \`${change?.old}\` → \`${change?.new}\``);
    }
    lines.push("");
  }

  const setPaths = Object.keys(diff.sets_changed).sort();
  if (setPaths.length > 0) {
    lines.push("## Named-set changes\n");
    for (const setPath of setPaths) {
      const change = diff.sets_changed[setPath];
      if (!change) continue;
      lines.push(`### \`${setPath}\`\n`);
      if (change.added.length > 0) {
        const names = change.added.map((n) => `\`${n}\``).join(", ");
        lines.push(`- **Added** (${change.added.length}): ${names}`);
      }
      if (change.removed.length > 0) {
        const names = change.removed.map((n) => `\`${n}\``).join(", ");
        lines.push(`- **Removed** (${change.removed.length}): ${names}`);
      }
      lines.push("");
    }
  }

  const otherKeys = Object.keys(diff.other_changed).sort();
  if (otherKeys.length > 0) {
    lines.push("## Other top-level changes\n");
    for (const key of otherKeys) {
      lines.push(`- \`${key}\`: value differs`);
    }
    lines.push("");
  }

  if (metaKeys.length === 0 && setPaths.length === 0 && otherKeys.length === 0) {
    lines.push("_No differences detected._\n");
  }

  return lines.join("\n");
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

/**
 * Walk every dict in the object tree; emit the path to any node shaped like
 * `{ names: string[] }`. Mirrors the Python `_named_set_paths` helper. Keyed
 * by the dotted-string form of the path so duplicates de-dupe naturally.
 */
function namedSetPaths(obj: unknown, path: string[] = []): Map<string, string[]> {
  const out = new Map<string, string[]>();
  if (!isPlainObject(obj)) return out;
  if (Array.isArray(obj.names) && path.length > 0) {
    out.set(path.join("."), [...path]);
  }
  for (const [key, value] of Object.entries(obj)) {
    const sub = namedSetPaths(value, [...path, key]);
    for (const [k, v] of sub) {
      out.set(k, v);
    }
  }
  return out;
}

function getAt(obj: unknown, path: string[]): unknown {
  let cur: unknown = obj;
  for (const part of path) {
    if (!isPlainObject(cur) || !(part in cur)) {
      return null;
    }
    cur = cur[part];
  }
  return cur;
}

function extractNames(node: unknown): string[] {
  if (!isPlainObject(node)) return [];
  const names = node.names;
  if (!Array.isArray(names)) return [];
  return names.filter((n): n is string => typeof n === "string");
}

/**
 * Structural equality for JSON-shaped values. Mirrors Python `==` semantics
 * for the dict/list/scalar types we actually see in contracts.
 */
function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null) return a === b;
  if (typeof a !== typeof b) return false;
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!deepEqual(a[i], b[i])) return false;
    }
    return true;
  }
  if (isPlainObject(a) && isPlainObject(b)) {
    const ak = Object.keys(a);
    const bk = Object.keys(b);
    if (ak.length !== bk.length) return false;
    for (const k of ak) {
      if (!(k in b)) return false;
      if (!deepEqual(a[k], b[k])) return false;
    }
    return true;
  }
  return false;
}
